src/ray/raylet/node_manager.cc

```diff
@@ -336,6 +336,7 @@ void NodeManager::Heartbeat() {
       static_cast<int64_t>(now_ms - last_debug_dump_at_ms_) > debug_dump_period_) {
     DumpDebugState();
     RecordMetrics();
+    WarnResourceDeadlock();
```

Comment:
Let's call this in `DispatchTasks` instead, and only call it if we weren't able to dispatch any tasks. Otherwise, we'll end up pushing an error every heartbeat for as long as the deadlock is happening (which is probably forever).

Reply:
I tried doing this, but it ended up printing out too many false positives. The issue is that you only want to fire the warning if there has been a significant delay, and right after DispatchTasks is not it (though if you wait even a tiny bit of time resources could immediately free up, like if a task returns right after creating an actor).

I think printing forever is fine -- it is deadlocked after all.

Reply:
Hmm, maybe I'm misunderstanding how DispatchTasks works, but if DispatchTasks doesn't succeed in scheduling anything, and the conditions that you check are true (no running tasks), then doesn't that mean it'll never succeed again? I think when this happens, it can only be because there are no available workers or if all the cores are taken up by actors. We can make sure it's the second case if we also check that there are no resources available.

Reply:
Not necessarily, a block/unblock could free up resources, here is the example I was testing:
```
import ray

@ray.remote(num_cpus=1)
class A:
    def f(self): pass


@ray.remote
def f():
    a = A.remote()
    b = A.remote()
    c = A.remote()
    print("get 1")
    ray.get(a.f.remote())
    print("get 2")
    ray.get(b.f.remote())
    print("get 3")
    ray.get(c.f.remote())

ray.init(num_cpus=2)
ray.get(f.remote())
```